File: agent/eterion_agent/models/streaming.py
"""Normalize provider chunks into frontend-facing content deltas.

Thinking and tool-call chunks will get their own adapters when those execution
paths are implemented; raw provider payloads must never become the IM contract.
"""

from __future__ import annotations

from typing import Any


def extract_content_delta(chunk: Any) -> str:
    text = getattr(chunk, "text", None)
    if isinstance(text, str) and text:
        return text

    content = getattr(chunk, "content", "")
    if isinstance(content, str):
        return content
    if isinstance(content, list):
        return "".join(
            str(block.get("text", ""))
            for block in content
            if isinstance(block, dict) and block.get("type") in {"text", "output_text"}
        )
    return ""

File: agent/src/models/streaming.ts
/**
 * 不同 Provider 的 chunk 形状可能不同，这里只抽取正式 Content。
 * reasoning/tool-call block 不能误混进前端正式回复。
 */
export function extractContentDelta(chunk: unknown): string {
  if (!isRecord(chunk)) return '';

  if (typeof chunk.text === 'string' && chunk.text) {
    return chunk.text;
  }

  const content = chunk.content;
  if (typeof content === 'string') return content;
  if (!Array.isArray(content)) return '';

  return content
    .filter(isRecord)
    .filter((block) => block.type === 'text' || block.type === 'output_text')
    .map((block) => (typeof block.text === 'string' ? block.text : ''))
    .join('');
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null;
}
